File: module3_segmentation/create_segmentation_overlay.m
function overlayImg = create_segmentation_overlay(enhancedImg, results, mode)
% CREATE_SEGMENTATION_OVERLAY Generates an RGB uint8 image with annotated segmentation masks.
%
% Syntax:
%   overlayImg = create_segmentation_overlay(enhancedImg, results, mode)
%
% Inputs:
%   enhancedImg - RGB fundus image (uint8)
%   results     - Struct containing the logical masks:
%                   .vesselMask
%                   .opticDiscMask
%                   .foveaMask
%                   .exudateMask
%                   .microaneurysmMask
%                   .hemorrhageMask
%                   .vesselAnalysis (for tortuosity segments)
%   mode        - Visualization mode: 'normal' or 'tortuosity' (default: 'normal')
%
% Outputs:
%   overlayImg - RGB uint8 composite image with overlay

    if nargin < 3 || isempty(mode)
        mode = 'normal';
    end

    [H, W, C] = size(enhancedImg);
    
    % Ensure base image is RGB uint8
    if C == 1
        rgbBase = cat(3, enhancedImg, enhancedImg, enhancedImg);
    else
        rgbBase = enhancedImg;
    end
    
    % Initialize output as a double for easy blending operations
    rgbOut = double(rgbBase);
    
    % Define custom colors (RGB [0, 255])
    colorVesselsNormal = [0, 255, 0];       % Green
    colorVesselLowTort = [0, 255, 0];       % Green
    colorVesselModTort = [255, 128, 0];     % Orange
    colorVesselHighTort = [255, 0, 0];      % Red
    
    colorDiscBoundary = [0, 128, 255];      % Bright Blue
    colorFovea = [255, 0, 255];             % Purple/Magenta
    colorExudate = [255, 255, 0];           % Yellow
    colorMA = [255, 165, 0];                % Orange (Microaneurysm)
    colorHemorrhage = [255, 0, 0];          % Red (Hemorrhage)
    
    alpha = 0.5; % Blending factor for solid overlays
    
    if strcmp(mode, 'tortuosity')
        % TORTUOSITY MODE: Color-code vessels by segment tortuosity
        % Get segment information from vessel analysis
        if isfield(results, 'vesselAnalysis') && isfield(results.vesselAnalysis, 'segments')
            segments = results.vesselAnalysis.segments;
            
            % Draw unsegmented skeleton pixels in dim green first
            skelMask = results.vesselSkeleton;
            for ch = 1:3
                layer = rgbOut(:, :, ch);
                layer(skelMask) = 100 * (ch == 2); % Dim green for background skeleton
                rgbOut(:, :, ch) = layer;
            end
            
            % Overlay segments colored by tortuosity
            for idx = 1:numel(segments)
                seg = segments(idx);
                pixelIdxs = seg.pixels;
                
                % Determine color based on segment tortuosity
                if seg.tortuosity <= 1.10
                    col = colorVesselLowTort;
                elseif seg.tortuosity <= 1.25
                    col = colorVesselModTort;
                else
                    col = colorVesselHighTort;
                end
                
                % Dilate the segment pixel indexes for better visibility in overlay
                segMask = false(H, W);
                segMask(pixelIdxs) = true;
                segMaskExpanded = imdilate(segMask, strel('disk', 1));
                expandedIdxs = find(segMaskExpanded);
                
                for ch = 1:3
                    layer = rgbOut(:, :, ch);
                    layer(expandedIdxs) = col(ch);
                    rgbOut(:, :, ch) = layer;
                end
            end
        else
            % Fallback if vessel analysis details are missing
            vMask = results.vesselMask;
            for ch = 1:3
                layer = rgbOut(:, :, ch);
                layer(vMask) = colorVesselsNormal(ch);
                rgbOut(:, :, ch) = layer;
            end
        end
    else
        % NORMAL MODE: Solid color segmentation overlay
        % 1. Vessels (Green)
        vMask = results.vesselMask;
        for ch = 1:3
            layer = rgbOut(:, :, ch);
            layer(vMask) = (1-alpha) * layer(vMask) + alpha * colorVesselsNormal(ch);
            rgbOut(:, :, ch) = layer;
        end
    end
    
    % 2. Optic Disc Boundary (Blue)
    if isfield(results, 'opticDiscMask') && sum(results.opticDiscMask(:)) > 0
        % Find boundary of the optic disc using morphological gradient
        discBoundary = imdilate(results.opticDiscMask, strel('disk', 1)) & ~results.opticDiscMask;
        % Make the boundary line thicker for visualization
        discBoundary = imdilate(discBoundary, strel('disk', 2));
        
        for ch = 1:3
            layer = rgbOut(:, :, ch);
            layer(discBoundary) = colorDiscBoundary(ch);
            rgbOut(:, :, ch) = layer;
        end
    end
    
    % 3. Exudates (Yellow)
    if isfield(results, 'exudateMask') && sum(results.exudateMask(:)) > 0
        exMask = results.exudateMask;
        % Slightly dilate exudate candidates to make them visible
        exMaskVis = imdilate(exMask, strel('disk', 1));
        for ch = 1:3
            layer = rgbOut(:, :, ch);
            layer(exMaskVis) = (1-alpha) * layer(exMaskVis) + alpha * colorExudate(ch);
            rgbOut(:, :, ch) = layer;
        end
    end
    
    % 4. Hemorrhages (Red)
    if isfield(results, 'hemorrhageMask') && sum(results.hemorrhageMask(:)) > 0
        heMask = results.hemorrhageMask;
        heMaskVis = imdilate(heMask, strel('disk', 1));
        for ch = 1:3
            layer = rgbOut(:, :, ch);
            layer(heMaskVis) = (1-alpha) * layer(heMaskVis) + alpha * colorHemorrhage(ch);
            rgbOut(:, :, ch) = layer;
        end
    end
    
    % 5. Microaneurysms (Orange)
    if isfield(results, 'microaneurysmMask') && sum(results.microaneurysmMask(:)) > 0
        maMask = results.microaneurysmMask;
        % Microaneurysms are very small, dilate them to form visible dots
        maMaskVis = imdilate(maMask, strel('disk', 2));
        for ch = 1:3
            layer = rgbOut(:, :, ch);
            layer(maMaskVis) = colorMA(ch);
            rgbOut(:, :, ch) = layer;
        end
    end
    
    % 6. Fovea Location (Purple)
    if isfield(results, 'foveaMask') && sum(results.foveaMask(:)) > 0
        % Draw an outer crosshair marker for fovea position
        foveaBoundary = imdilate(results.foveaMask, strel('disk', 1)) & ~results.foveaMask;
        foveaBoundary = imdilate(foveaBoundary, strel('disk', 1));
        
        for ch = 1:3
            layer = rgbOut(:, :, ch);
            layer(foveaBoundary) = colorFovea(ch);
            rgbOut(:, :, ch) = layer;
        end
    end
    
    % Convert final blended double matrix back to uint8 RGB
    overlayImg = uint8(min(255, max(0, rgbOut)));
end

File: module3_segmentation python/module3_segmentation python/create_segmentation_overlay.py
"""
Module 3: Segmentation Overlay Generator (Python Implementation)
SIH 2026 - Problem Statement SIH260038
"""

import numpy as np

def create_segmentation_overlay(image, results, style='normal'):
    """
    Creates an RGB composite visualization overlay of retinal structures and lesions.
    """
    if len(image.shape) == 2:
        overlay = np.stack([image, image, image], axis=-1)
    else:
        overlay = image.copy()
        
    alpha = 0.55
    
    # 1. Vessels (Green: [0, 255, 0])
    if 'vesselMask' in results and results['vesselMask'] is not None:
        mask = results['vesselMask']
        overlay[mask, 0] = (1 - alpha) * overlay[mask, 0] + alpha * 0
        overlay[mask, 1] = (1 - alpha) * overlay[mask, 1] + alpha * 255
        overlay[mask, 2] = (1 - alpha) * overlay[mask, 2] + alpha * 0
        
    # 2. Optic Disc (Blue: [0, 180, 255])
    if 'opticDiscMask' in results and results['opticDiscMask'] is not None:
        mask = results['opticDiscMask']
        overlay[mask, 0] = (1 - alpha) * overlay[mask, 0] + alpha * 0
        overlay[mask, 1] = (1 - alpha) * overlay[mask, 1] + alpha * 180
        overlay[mask, 2] = (1 - alpha) * overlay[mask, 2] + alpha * 255
        
    # 3. Optic Cup (Cyan: [0, 255, 255])
    if 'opticCupMask' in results and results['opticCupMask'] is not None:
        mask = results['opticCupMask']
        overlay[mask, 0] = (1 - alpha) * overlay[mask, 0] + alpha * 0
        overlay[mask, 1] = (1 - alpha) * overlay[mask, 1] + alpha * 255
        overlay[mask, 2] = (1 - alpha) * overlay[mask, 2] + alpha * 255
        
    # 4. Fovea (Purple: [200, 50, 255])
    if 'foveaMask' in results and results['foveaMask'] is not None:
        mask = results['foveaMask']
        overlay[mask, 0] = (1 - alpha) * overlay[mask, 0] + alpha * 200
        overlay[mask, 1] = (1 - alpha) * overlay[mask, 1] + alpha * 50
        overlay[mask, 2] = (1 - alpha) * overlay[mask, 2] + alpha * 255
        
    # 5. Exudates (Yellow: [255, 255, 0])
    if 'exudateMask' in results and results['exudateMask'] is not None:
        mask = results['exudateMask']
        overlay[mask, 0] = 255
        overlay[mask, 1] = 255
        overlay[mask, 2] = 0
        
    # 6. Microaneurysms (Orange: [255, 140, 0])
    if 'microaneurysmMask' in results and results['microaneurysmMask'] is not None:
        mask = results['microaneurysmMask']
        overlay[mask, 0] = 255
        overlay[mask, 1] = 140
        overlay[mask, 2] = 0
        
    # 7. Hemorrhages (Red: [255, 0, 0])
    if 'hemorrhageMask' in results and results['hemorrhageMask'] is not None:
        mask = results['hemorrhageMask']
        overlay[mask, 0] = 255
        overlay[mask, 1] = 0
        overlay[mask, 2] = 0
        
    return overlay.astype(np.uint8)
